From cdd10398e71a1843ef99ed545bbb872b6cb9d249 Mon Sep 17 00:00:00 2001
From: Namjae Jeon <namjae.jeon@samsung.com>
Date: Wed, 26 May 2021 15:22:37 +0900
Subject: [PATCH] cifsd: add goto fail in asn1_oid_decode()
Git-commit: cdd10398e71a1843ef99ed545bbb872b6cb9d249
Patch-mainline: v5.15-rc1
References: bsn#12
Modified-by-SEL: No


Add goto fail in asn1_oid_decode() to clean-up exception handling code.

Reviewed-by: Dan Carpenter <dan.carpenter@oracle.com>
Signed-off-by: Namjae Jeon <namjae.jeon@samsung.com>
Signed-off-by: Steve French <stfrench@microsoft.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/cifsd/asn1.c | 26 +++++++++++---------------
 1 file changed, 11 insertions(+), 15 deletions(-)

diff --git a/fs/cifsd/asn1.c b/fs/cifsd/asn1.c
index aa6ea855c422..f2628dc3490f 100644
--- a/fs/cifsd/asn1.c
+++ b/fs/cifsd/asn1.c
@@ -74,11 +74,8 @@ static bool asn1_oid_decode(const unsigned char *value, size_t vlen,
 
 	optr = *oid;
 
-	if (!asn1_subid_decode(&iptr, end, &subid)) {
-		kfree(*oid);
-		*oid = NULL;
-		return false;
-	}
+	if (!asn1_subid_decode(&iptr, end, &subid))
+		goto fail;
 
 	if (subid < 40) {
 		optr[0] = 0;
@@ -95,19 +92,18 @@ static bool asn1_oid_decode(const unsigned char *value, size_t vlen,
 	optr += 2;
 
 	while (iptr < end) {
-		if (++(*oidlen) > vlen) {
-			kfree(*oid);
-			*oid = NULL;
-			return false;
-		}
+		if (++(*oidlen) > vlen)
+			goto fail;
 
-		if (!asn1_subid_decode(&iptr, end, optr++)) {
-			kfree(*oid);
-			*oid = NULL;
-			return false;
-		}
+		if (!asn1_subid_decode(&iptr, end, optr++))
+			goto fail;
 	}
 	return true;
+
+fail:
+	kfree(*oid);
+	*oid = NULL;
+	return false;
 }
 
 static bool
-- 
2.33.0

